src/core/server/saved_objects/service/lib/update_objects_spaces.ts

```diff
@@ -0,0 +1,323 @@
+  objects: SavedObjectsUpdateObjectsSpacesResponseObject[];
+}
+
+/**
+ *
+ * @public
+ */
+export interface SavedObjectsUpdateObjectsSpacesResponseObject {
+  /** The type of the referenced object */
+  type: string;
+  /** The ID of the referenced object */
+  id: string;
+  /** The space(s) that the referenced object exists in */
+  spaces: string[];
+  /** Included if there was an error updating this object's spaces */
+  error?: SavedObjectError;
+}
+
+// eslint-disable-next-line @typescript-eslint/consistent-type-definitions
+type Left = { tag: 'Left'; error: SavedObjectsUpdateObjectsSpacesResponseObject };
+// eslint-disable-next-line @typescript-eslint/consistent-type-definitions
+type Right = { tag: 'Right'; value: Record<string, any> };
+type Either = Left | Right;
+const isLeft = (either: Either): either is Left => either.tag === 'Left';
+const isRight = (either: Either): either is Right => either.tag === 'Right';
+
+export interface UpdateObjectsSpacesParams {
+  registry: ISavedObjectTypeRegistry;
+  allowedTypes: string[];
+  client: RepositoryEsClient;
+  serializer: SavedObjectsSerializer;
+  getIndexForType: (type: string) => string;
+  collectMultiNamespaceReferences: (
+    objects: SavedObjectsCollectMultiNamespaceReferencesObject[]
+  ) => Promise<SavedObjectsCollectMultiNamespaceReferencesResponse>;
+  objects: SavedObjectsUpdateObjectsSpacesObject[];
+  spacesToAdd: string[];
+  spacesToRemove: string[];
+  options?: SavedObjectsUpdateObjectsSpacesOptions;
+}
+
+/**
+ * Gets all references and transitive references of the given objects. Ignores any object and/or reference that is not a multi-namespace
+ * type.
+ */
+export async function updateObjectsSpaces({
+  registry,
+  allowedTypes,
+  client,
+  serializer,
+  getIndexForType,
+  collectMultiNamespaceReferences,
+  objects,
+  spacesToAdd,
+  spacesToRemove,
+  options = {},
+}: UpdateObjectsSpacesParams): Promise<SavedObjectsUpdateObjectsSpacesResponse> {
+  if (!spacesToAdd.length && !spacesToRemove.length) {
+    throw SavedObjectsErrorHelpers.createBadRequestError(
+      'spacesToAdd and/or spacesToRemove must be a non-empty array of strings'
+    );
+  }
+  if (doArraysIntersect(spacesToAdd, spacesToRemove)) {
+    throw SavedObjectsErrorHelpers.createBadRequestError(
+      'spacesToAdd and spacesToRemove cannot contain any of the same strings'
+    );
+  }
+
+  const { namespace, includeReferences } = options;
+  const expandedObjects: Array<
+    SavedObjectsUpdateObjectsSpacesObject & { isMissing?: boolean }
+  > = includeReferences ? (await collectMultiNamespaceReferences(objects)).objects : objects;
+
+  let bulkGetRequestIndexCounter = 0;
+  const expectedBulkGetResults: Either[] = expandedObjects.map((object) => {
+    const { type, id, spaces, version, isMissing } = object;
+
+    if (!allowedTypes.includes(type)) {
+      const error = errorContent(SavedObjectsErrorHelpers.createGenericNotFoundError(type, id));
```

Comment:
Disallowed types are treated inconsistently in the repository. Some methods treat it as Not Found and/or use `createGenericNotFoundError` (find, delete, get, resolve, update, bulkUpdate, addToNamespaces, removeFromNamespaces) and others use `createUnsupportedTypeError` (create, bulkCreate, bulkGet, checkConflicts, incrementCounter). Since this is effectively an update operation, I modeled it after update/bulkUpdate. I think there's room for improvement in the SOR in general with regards to this.

Reply:
Fair enough. If there isn't a legitimate reason for this inconsistency, then my vote is to change to `createUnsupportedTypeError`, but I'll defer to you. I don't have a strong enough preference to argue this nit 🙂

Reply:
I'll leave it as-is for now and raise it as a question for the Core team.